File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import defaultdict


def dfs(now, graph, result):
    while graph[now]:
        dfs(graph[now].pop(), graph, result)
    
    result.append(now)
    return result


def solution(tickets):
    graph = defaultdict(list)
    for start, end in tickets:
        graph[start].append(end)
        
    for key in graph:
        graph[key].sort(reverse=True)
        
    result = []
    dfs('ICN', graph, result)
    
    return result[::-1]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {

    private Map<String, PriorityQueue<String>> graph;
    private List<String> result;

    public String[] solution(String[][] tickets) {
        graph = new HashMap<>();
        result = new ArrayList<>();

        for (String[] ticket : tickets) {
            graph.computeIfAbsent(ticket[0], k -> new PriorityQueue<>())
                 .offer(ticket[1]);
        }

        dfs("ICN");

        Collections.reverse(result);
        return result.toArray(new String[0]);
    }

    private void dfs(String now) {
        PriorityQueue<String> pq = graph.get(now);
        while (pq != null && !pq.isEmpty()) {
            dfs(pq.poll());
        }
        result.add(now);
    }
}